cu29_derive/src/lib.rs

```diff
@@ -397,15 +401,19 @@ fn extract_tasks_types(copper_config: &CuConfig) -> (Vec<String>, Vec<Type>) {
     (all_types_names, all_types)
 }
 
-fn extract_msg_types(runtime_plan: &Vec<CuExecutionStep>) -> Vec<Type> {
+fn extract_msg_types(runtime_plan: &CuExecutionLoop) -> Vec<Type> {
     runtime_plan
+        .steps
         .iter()
-        .filter_map(|step| {
-            if step.output_msg_type.is_none() {
-                None
-            } else {
-                Some(parse_str::<Type>(step.output_msg_type.clone().unwrap().as_str()).unwrap())
+        .filter_map(|unit| match unit {
+            CuExecutionUnit::Step(step) => {
+                if step.output_msg_type.is_none() {
+                    None
+                } else {
+                    Some(parse_str::<Type>(step.output_msg_type.clone().unwrap().as_str()).unwrap())
+                }
             }
+            CuExecutionUnit::Loop(_) => todo!("Needs to be implemented"),
```

Comment:
Is it going to go through all of the steps in the loop and aggregate all of the message types? 

And do we care about duplicates in this return type? Should it be changed into a set or is the type vec being indexed into or something?

Reply:
I believe it works as is ...  we will need an array of the type instead of just a type for each step of the loop as output. And at runtime we need to generate the index for the loop.
in the vector it will be:
[mainloop_type1 [loop1_type1[] loop1type2[]] mainloop_type2 ...] 

we might want to go through some design lol